src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,397 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Runtime.Internal.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    public static class ReadOnlyMemoryUtils
+    {
+
+        /// <summary>
+        /// Compares two <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> byte by byte.
+        /// </summary>
+        public static bool Equals(ReadOnlyMemory<char> b, ReadOnlyMemory<char> a)
+        {
+            if (a.Length != b.Length)
+                return false;
+
+            Contracts.Assert(a.IsEmpty == b.IsEmpty);
+            var aSpan = a.Span;
+            var bSpan = b.Span;
+
+            for (int i = 0; i < a.Length; i++)
+            {
+                if (aSpan[i] != bSpan[i])
+                    return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Compare equality with the given system string value.
+        /// </summary>
+        public static bool EqualsStr(string s, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValueOrNull(s);
+
+            if (s == null)
+                return memory.Length == 0;
+
+            if (s.Length != memory.Length)
+                return false;
+
+            var span = memory.Span;
+            for (int i = 0; i < memory.Length; i++)
+            {
+                if (s[i] != span[i])
```

Comment:
Should use `MemoryExtensions.SequenceEquals` instead.

Reply:
That's totally fine. You can easily convert a string to Span using `.AsSpan()`. If we targeted just .NET Core 2.1, there is an implicit conversion. But since we are only .NET Standard 2.0, you need to call the `.AsSpan()` extension method.

The "Performance" section of [this blog post](https://blogs.msdn.microsoft.com/dotnet/2018/08/20/bing-com-runs-on-net-core-2-1/) explains why you want to call these methods - because they are vectorized.

Reply:
Awesome!!!

---
In reply to: [216850011](https://github.com/dotnet/machinelearning/pull/863#discussion_r216850011) [](ancestors = 216850011)